### ISSUE TITLE
Replication support

### ISSUE THREAD

**Issue by dvarrazzo** [psycopg/psycopg#71]
Add support for streaming replication as in psycopg2

**Comment by sassemlal** [comment]
Hi, 

If this still an issue, I could take a look and use your guidance to implement this.

**Comment by dufferzafar** [comment]
@dvarrazzo Could you please also explain exactly what would need to be added to support replication directly in psycopg3? 

I'm taking a look at the code right now & having some pointers on exactly what needs to be done would be great!

**Comment by dvarrazzo** [comment]
Replication support is not implemented in psycopg 3. It needs complete redesign and implementation. If you want to contribute its implementation you are welcome, but it is a major work.

**Comment by dufferzafar** [comment]
Yeah, I realised that after looking at the code of pg2 & pg3. I guess it'll require a lot of interfacing with pq as well?

My current plan is to use something like [`wal2json`](https://github.com/eulerto/wal2json/) extension on the server, then create a slot and get_changes in the JSON format directly from the server. 

Or perhaps I won't do this in async mode and just use pg2 directly. I found pypgoutput which does client side decoding of pgoutput format to JSON: https://github.com/dgea005/pypgoutput

**Comment by sergiors** [comment]
👀

**Comment by RPG-fan** [comment]
## Working Implementation Available

I needed logical replication for a project and built a working solution. Since others seem to be stuck on this, I'm sharing it.

**Gist:** [https://gist.github.com/RPG-fan/b6d578e45712ae9467b05d6ac4e8dbc6]

### How it works

The implementation uses a "bridge" pattern:
1. Uses psycopg3 to establish connection and send START_REPLICATION
2. After entering COPY_BOTH mode, hands off to raw socket I/O
3. Reads/writes PostgreSQL wire protocol messages directly
4. Includes complete pgoutput binary protocol parser

### Why this approach

psycopg3's state machine gets confused trying to handle COPY protocol for replication. By taking direct control of the socket after COPY_BOTH is established, we bypass the state checking while still using psycopg3's solid connection foundation.

### Status

Running in production with 9 concurrent replication consumers. Handles BEGIN, COMMIT, RELATION, INSERT, UPDATE, DELETE, and TRUNCATE messages. Properly acknowledges LSN positions and handles keepalives.

### Usage

```
stream = LogicalReplicationStream(
    conn=conn,  # AsyncConnection with replication=database
    slot_name="my_slot",
    publication_name="my_pub"
)

async for raw_msg, decoded_msg in stream.start_replication():
    if isinstance(decoded_msg, InsertMessage):
        # Handle insert
        pass
    await stream.send_feedback(raw_msg.data_end)
```

No promises of long-term support, but it works for me and might help others who are blocked. Feel free to use/adapt.

**Comment by dvarrazzo** [comment]
@RPG-fan thank you, it looks interesting.

I was able to run your proof of concept ok. It didn't work with Postgres 15 but it worked ok on Postgres 17. If anyone wants to play with it they might need a bit of setup:

```
# edit postgresql.conf
wal_level = logical
max_wal_senders = 10
max_replication_slots = 10
# restart the server
```

In a psql shell:

```
SELECT * FROM pg_create_logical_replication_slot('my_replication_slot', 'pgoutput');
CREATE PUBLICATION my_publication FOR ALL TABLES;
```

Then the script should be more or less what expected.

The modification for psycopg to accept the command are pretty minimal:
```diff
diff --git a/psycopg/psycopg/_cursor_base.py b/psycopg/psycopg/_cursor_base.py
index 10ddc1f3..fa1ad59e 100644
--- a/psycopg/psycopg/_cursor_base.py
+++ b/psycopg/psycopg/_cursor_base.py
@@ -465,7 +465,12 @@ class BaseCursor(Generic[ConnectionType, Row]):
 
         for res in results:
             status = res.status
-            if status != TUPLES_OK and status != COMMAND_OK and status != EMPTY_QUERY:
+            if (
+                status != TUPLES_OK
+                and status != COMMAND_OK
+                and status != EMPTY_QUERY
+                and status != COPY_BOTH
+            ):
                 self._raise_for_result(res)
 
     def _raise_for_result(self, result: PGresult) -> NoReturn:
```

This is enough to allow psycopg to run `conn.execute("START_REPLICATION ...")` instead of using the lower level libpq. Of course from there on the connection will be broken for further usage.

Would you like to propose a MR to add logical replication support to psycopg or would you just rather leave it as it is?

Thank you again for the demo!

**Comment by RPG-fan** [comment]
Thank you for helping explain how to use this. It's a basic implementation right now, but I'll likely be expanding upon it in the near(ish) future to increase the flexibility. I can take a deeper look at how to properly integrate it when I get some more time for a potential merge, but I wanted to at least get this version out there in case someone else would find it useful.

**Comment by RPG-fan** [comment]
Added some more today to make it more reliable/stable:
UPDATE (October 15, 2025): Added production stability features:
- Reconnection logic with exponential backoff
- LSN tracking across reconnections
- Keepalive handling (both primary and unsolicited)
- Proper initialization handshake
- Graceful connection cleanup
https://gist.github.com/RPG-fan/b6d578e45712ae9467b05d6ac4e8dbc6

**Comment by dvarrazzo** [comment]
In #1219 I am committing the change described above, allowing to run START_REPLICATION without errors. It will be released in Psycopg 3.3 (release is imminent).

No further replication support is planned for release 3.3.

**Comment by DylanYoung** [comment]
FYI: I've thrown up a POC in #1311.  It still needs a bit of work, but it's close and fairly feature-complete.   I think I broke the existing tests with my pytest fixtures, so I'll have to look at that before CI runs properly.

**Comment by dvarrazzo** [comment]
Thank you @DylanYoung. However, as described in #1311, there is some ongoing work that goes in a direction different for this design.

We will be ready to give information about the state of the work on this and on other expected features in the near future.

**Comment by DylanYoung** [comment]
@dvarrazzo  Oki dokey.  Like I said, I'm happy to make modifications if desired.  I'm not sure how it could be much more advanced, but I'm excited to see what you have planned.
